lib/spack/spack/directory_layout.py

```diff
@@ -261,24 +257,8 @@ def ensure_installed(self, spec):
                 "  " + path)
 
         installed_spec = self.read_spec(spec_file_path)
-        if installed_spec == spec:
+        if installed_spec.dag_hash() == spec.dag_hash():
```

Comment:
Slight simplification:
```suggestion
        if installed_spec.dag_hash() != spec.dag_hash():
            raise ...
```
If we do so we can avoid the `else` and explicit return